fdbserver/ptxn/TLogServer.actors.cpp

```diff
@@ -0,0 +1,1592 @@
+	TraceEvent("TLogCheckEmptyQueueBegin", self->dbgid);
+	try {
+		bool recoveryFinished = wait(self->persistentQueue->initializeRecovery(0));
+		if (recoveryFinished)
+			return Void();
+		TLogQueueEntry r = wait(self->persistentQueue->readNext(self.getPtr()));
+		throw internal_error();
+	} catch (Error& e) {
+		if (e.code() != error_code_end_of_stream)
+			throw;
+		TraceEvent("TLogCheckEmptyQueueEnd", self->dbgid);
+		return Void();
+	}
+}
+
+ACTOR Future<Void> checkRecovered(Reference<TLogServerData> self) {
+	TraceEvent("TLogCheckRecoveredBegin", self->dbgid);
+	Optional<Value> v = wait(self->persistentData->readValue(StringRef()));
+	TraceEvent("TLogCheckRecoveredEnd", self->dbgid);
+	return Void();
+}
+
+bool tlogTerminated(Reference<TLogTeamData> self,
+                    IKeyValueStore* persistentData,
+                    TLogQueue* persistentQueue,
+                    Error const& e) {
+	// Dispose the IKVS (destroying its data permanently) only if this shutdown is definitely permanent.  Otherwise just
+	// close it.
+	if (e.code() == error_code_worker_removed || e.code() == error_code_recruitment_failed) {
+		persistentData->dispose();
+		persistentQueue->dispose();
+	} else {
+		persistentData->close();
+		persistentQueue->close();
+	}
+
+	if (e.code() == error_code_worker_removed || e.code() == error_code_recruitment_failed ||
+	    e.code() == error_code_file_not_found) {
+		TraceEvent("TLogTerminated", self->dbgid).error(e, true);
+		return true;
+	} else
+		return false;
+}
+
+void stopAllTLogs(Reference<TLogServerData> self, UID newLogId) {
+	for (auto& team : self->tLogTeams) {
+		for (auto it : team.second->id_data) {
+			if (!it.second->stopped) {
+				TraceEvent("TLogStoppedByNewRecruitment", self->dbgid)
+				    .detail("LogId", it.second->logId)
+				    .detail("StoppedId", it.first.toString())
+				    .detail("RecruitedId", newLogId)
+				    .detail("EndEpoch", it.second->logSystem->get().getPtr() != 0);
+				if (!it.second->isPrimary && it.second->logSystem->get()) {
+					it.second->removed = it.second->removed && it.second->logSystem->get()->endEpoch();
+				}
+				if (it.second->committingQueue.canBeSet()) {
+					it.second->committingQueue.sendError(worker_removed());
+				}
+			}
+			it.second->stopped = true;
+			if (!it.second->recoveryComplete.isSet()) {
+				it.second->recoveryComplete.sendError(end_of_stream());
+			}
+			it.second->stopCommit.trigger();
+		}
+	}
+}
+
+// Initialize a new tLog team (if !recoverFrom.size()) or restore from network
+// persistentData: All teams share single persistentData, in-memory for now.
+// persistentQueue: dummy queue for now, we will let the team find out
+ACTOR Future<Void> tLogTeamRecovery(Reference<TLogTeamData> self, Promise<Void> recovered) {
+	try {
+		wait(ioTimeoutError(checkEmptyQueue(self) && checkRecovered(self), SERVER_KNOBS->TLOG_MAX_CREATE_DURATION));
+
+		// Disk errors need a chance to kill this actor.
```

Comment:
Can you elaborate this comment a bit further?

Reply:
I think wait(delay(0.000001)) here is to yield the event loop so that disk error can first takes over the event loop and kills this actor. Not sure though. It comes from old code.